### PR TITLE
feat: add dependency and env tests

### PR DESCRIPTION
**Pull request by keenahn**
Initial add.

To test, add this as a dependency on another project and run the tests (see README.md).

The purpose of this is:
1. DRY up these tests, which should be in every invisible project
2. Enforce the existence of `env.sample` and DRY it with the test

The last thing is, this package includes a few changes that I want to re-merge into the inv-project-template

1. Updated templates for Changelog, Readme, and circle.yml
2. `posttest.sh`
3. Updates to the package.json

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# basic-tests
+
+> Basic tests for your project: requiring each file and testing environment variables
```
Comment (gunar) [inline README.md]: Lacking full stop.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+## Prerequisites
+
+This package requires Mocha. If you want to use it for other testing frameworks, make an Issue :).
```
Comment (gunar) [inline README.md]: "Open" an Issue?

File: README.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+## Install
+
+```bash
+yarn add --dev @invisible/basic-tests
```
Comment (gunar) [inline README.md]: I prefer to use `npm install` for such installation guides, but ok.

File: README.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+```
+
+## Use
```
Comment (gunar) [inline README.md]: Usage?

File: README.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+This module exposes two methods `dependency` and `env`.
+
+1. `dependency()` will run mocha tests to require every file in your project, aside from those ignored by your `.gitignore` or are in your `test` directory. The reason why this test is useful is that it ensures that you don't have any obvious errors in your files. It should also encourage you to not have any side effects on requiring files.
+
+2. `env()` runs mocha tests to ensure that all env vars in `env.sample` are actually set. This is useful so that you won't try to run your package when it is missing a required environment variable.
+
```
Comment (gunar) [inline README.md]: :+1:

File: README.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+```javascript
+
+// test/index.js
+
+'use strict'
+
+const { dependency, env } = require('@invisible/basic-tests')
+
+dependency()
+env()
```
Comment (gunar) [inline README.md]: Why don't we just expose this as a binary?
Make mocha a dependency here, which would uncouple the present package's tests with the consumer's.

File: circle.yml [hunk 7]
```diff
@@ -0,0 +1,26 @@
+dependencies:
+  pre:
+    # Install yarn
+    - curl -o- -L https://yarnpkg.com/install.sh | bash
```
Comment (gunar) [inline circle.yml]: So you prefer this over making yarn a devDependency?

File: package.json [hunk 8]
```diff
@@ -0,0 +1,47 @@
+{
+  "name": "@invisible/basic-tests",
+  "license": "MIT",
+  "version": "1.0.0",
+  "description": "Basic tests for your project: requiring each file and testing environment variables",
```
Comment (gunar) [inline package.json]: If we add more tests I'm sure we'll forget to update this.
Get rid of everything after the colon.

File: package.json [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    "name": "Invisible Technologies",
+    "url": "https://github.com/invisible-tech"
+  },
+  "main": "index.js",
+  "keywords": [],
+  "devDependencies": {
+    "@invisible/changelog-update": "^1.0.7",
+    "@invisible/eslint-config": "^1.1.6",
+    "@invisible/publish": "^1.1.8",
+    "app-root-dir": "^1.0.2",
```
Comment (gunar) [inline package.json]: This is probably a `dependency` not `devDependency`.

File: package.json [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    "eslint": "^4.6.1",
+    "espower-loader": "^1.2.2",
+    "glob-gitignore": "^1.0.5",
+    "ignore": "^3.3.5",
+    "mocha": "^3.5.3",
+    "nsp": "^2.8.0",
+    "nyc": "^11.2.1",
+    "power-assert": "^1.4.4",
+    "yarn": "^1"
```
Comment (gunar) [inline package.json]: I don't think this is necessary as you added it to circle.yml

File: test/env.js [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+    if (fs.existsSync(envSamplePath)) {
+      const str = fs.readFileSync(envSamplePath)
+      const config = dotenv.parse(str)
+      forEach(name => {
+        it(name, () => {
+          const actual = get(name)(process.env)
+          assert.notEqual(actual, undefined, `${name} is not set`)
+        })
+      })(keys(config))
```
Comment (gunar) [inline test/env.js]: Nice

**Comment by gunar** [comment]
I would like you to seriously consider exporting a binary.

File: README.md (hunk 1 above)
Comment (keenahn) [reply README.md]: 👎 It's not a sentence :)

File: README.md (hunk 2 above)
Comment (keenahn) [reply README.md]: 👍

File: README.md (hunk 3 above)
Comment (keenahn) [reply README.md]: nevar

File: README.md (hunk 4 above)
Comment (keenahn) [reply README.md]: 👍

File: circle.yml (hunk 7 above)
Comment (keenahn) [reply circle.yml]: No because you need `yarn` to install the packages in the first place.

**Comment by keenahn** [comment]
Re: exporting binaries

There's pros and cons to this. The con is that your `test` script in `package.json` would get more complicated. And then if someone wanted to use it, but they weren't using mocha, they would get mocha as a dependency.

I'll think on it

File: circle.yml (hunk 7 above)
Comment (gunar) [reply circle.yml]: How come Gear tests work without this, though?

**Comment by keenahn** [comment]
@gunar OK, I completely refactored this and squashed it to one commit.

1. Use `check-env` which cut a few lines (but added a dependency, I’m not sure if it was actually worth it, but whatever)
2. Just expose the two methods instead of forcing the user to use Mocha, but provide an example test file.
3. Use those two methods in a nice wrapped `bin` file (to remove the useless stack trace if it throws, but still exit with a non-zero value)

**Comment by keenahn** [comment]
Also, I'm not sure what should be in `dependencies` vs `devDependencies` I'm thinking

**Pull request closed without merging**